File: ui/shared/events.ts
/**
 * Tauri `listen()` 事件订阅封装（对应 runtime/src/events.rs 中的事件常量）。
 */

import { listen } from '@tauri-apps/api/event';
import type {
  DmState,
  EvokeScore,
  HistoryEntry,
  OperationProgress,
} from './types';

/** 取消订阅函数类型。 */
export type Unlisten = () => void;

/** 订阅状态变化事件（"state-changed"）。HudWindow 用于切换黄/绿灯。 */
export function onStateChanged(callback: (state: DmState) => void): Promise<Unlisten> {
  return listen<DmState>('state-changed', (event) => callback(event.payload));
}

/** 订阅历史记录新增事件（"history-updated"）。History 二级页用于实时刷新列表。 */
export function onHistoryUpdated(callback: (entry: HistoryEntry) => void): Promise<Unlisten> {
  return listen<HistoryEntry>('history-updated', (event) => callback(event.payload));
}

/** 订阅当前输入设备的实时音量（"input-level"，范围 0-1）。 */
export function onInputLevel(callback: (level: number) => void): Promise<Unlisten> {
  return listen<number>('input-level', (event) => callback(event.payload));
}

export function onEvokeScore(callback: (score: EvokeScore) => void): Promise<Unlisten> {
  return listen<EvokeScore>('evoke-score', (event) => callback(event.payload));
}

export function onOperationProgress(
  callback: (progress: OperationProgress) => void,
): Promise<Unlisten> {
  return listen<OperationProgress>('operation-progress', (event) => callback(event.payload));
}

export function onSettingsChanged(callback: (generation: number) => void): Promise<Unlisten> {
  return listen<{ generation: number }>('settings-changed', (event) => {
    callback(event.payload.generation);
  });
}

File: runtime/src/events.rs
//! Tauri 事件层：Runtime 向前端（MainWindow / HudWindow）广播的事件（见 brainstrom/plan.md §7）。

use tauri::{AppHandle, Emitter};

use crate::scoring::EvokeScore;
use crate::state_machine::State;
use crate::storage::HistoryEntry;

/// 事件名常量，供前端 `listen()` 订阅（见 ui/shared/events.ts）。
pub const EVENT_STATE_CHANGED: &str = "state-changed";
pub const EVENT_HISTORY_UPDATED: &str = "history-updated";
pub const EVENT_INPUT_LEVEL: &str = "input-level";
pub const EVENT_EVOKE_SCORE: &str = "evoke-score";

/// 状态变化时广播（HudWindow 据此切换黄/绿灯，MainWindow 首页据此更新提示）。
pub fn emit_state_changed(app: &AppHandle, state: State) -> Result<(), tauri::Error> {
    app.emit(EVENT_STATE_CHANGED, state)
}

/// 新增历史记录时广播（History 二级页据此实时刷新列表）。
pub fn emit_history_updated(app: &AppHandle, entry: HistoryEntry) -> Result<(), tauri::Error> {
    app.emit(EVENT_HISTORY_UPDATED, entry)
}

/// 输入设备页打开期间推送节流后的麦克风电平（0.0-1.0）。
pub fn emit_input_level(app: &AppHandle, level: f32) -> Result<(), tauri::Error> {
    app.emit(EVENT_INPUT_LEVEL, level.clamp(0.0, 1.0))
}

pub fn emit_evoke_score(app: &AppHandle, score: EvokeScore) -> Result<(), tauri::Error> {
    app.emit(EVENT_EVOKE_SCORE, score)
}
